File: BaekjoonAlgorithm/202012/BA_11724/BJAlgo_11724.py
from sys import stdin


# 일반적인 DFS 수행 (연결 요소 찾기)
def dfs(start):
    queue = [start]

    while queue:
        node = queue.pop()
        if not visited[node]:
            queue.extend(graph[node])
            visited[node] = True


n, m = map(int, stdin.readline().split())
graph = [[] for _ in range(1001)]
for _ in range(m):
    u, v = map(int, stdin.readline().split())
    # 무방향 그래프이기 때문에 양쪽 정점을 모두 저장해줍니다.
    graph[u].append(v)
    graph[v].append(u)

visited = [False] * 1001
count = 0
for i in range(1, n + 1):
    if not visited[i]:  # 아직 연결되지 않은 경우
        dfs(i)  # 연결 노드 찾기 수행
        count += 1
print(count)

File: BaekjoonAlgorithm/202012/BA_11724/BJAlgo_11724.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class BJAlgo_11724 {

    private static List<ArrayList<Integer>> graph;
    private static boolean[] visited;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        graph = new ArrayList<>();
        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        visited = new boolean[n + 1];
        int count = 0;

        for (int i = 1; i < n + 1; i++) {
            if (!visited[i]) {
                dfs(i);
                count += 1;
            }
        }

        System.out.println(count);
    }

    private static void dfs(int start) {
        visited[start] = true;

        for (int n : graph.get(start)) {
            if (!visited[n]) {
                dfs(n);
            }
        }
    }
}
